### PR TITLE
EA-1610: changed implementation of validate method

### PR DESCRIPTION
**Pull request by helinek**

### CODE REVIEW

File: apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java [hunk 1]
```diff
@@ -313,61 +315,29 @@ public ApikeyController(ApikeyRepo apikeyRepo) {
...
-        Date        now      = nowDtUtc.toDate();
-
-        if (StringUtils.isEmpty(api)) {
-            LOG.debug("no value for parameter 'api' supplied");
+    @PostMapping(path = "/validate")
+    public ResponseEntity<Apikey> validate(HttpServletRequest httpServletRequest) {
+        // When no apikey was supplied return 400
+        String id = getApikey(httpServletRequest);
+        if (null == id) {
+            LOG.debug(APIKEYMISSING);
```
Comment (gsergiu) [inline apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java]: the bad request response should also indicate the reason. (e.g. Authorization header required)

File: apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java [hunk 2]
```diff
@@ -313,61 +315,29 @@ public ApikeyController(ApikeyRepo apikeyRepo) {
...
+        LOG.debug("validate apikey: {}", id);
+
         // retrieve apikey & check if available
         Apikey apikey = this.apikeyRepo.findOne(id);
         if (null == apikey) {
-            LOG.debug(APIKEYNOTFOUND + " with value: " + id);
-            headers.add(APIKEYNOTFOUND, APIKEYNOTFOUND.toLowerCase());
-            return new ResponseEntity<>(headers, HttpStatus.NOT_FOUND);
+            LOG.debug(APIKEYNOTREGISTERED, id);
+            return new ResponseEntity<>(HttpStatus.UNAUTHORIZED);
```
Comment (gsergiu) [inline apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java]: the bad request response should also indicate the reason. (e.g. invalid api key)

File: apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java [hunk 3]
```diff
@@ -376,6 +346,8 @@ public ApikeyController(ApikeyRepo apikeyRepo) {
             return new ResponseEntity<>(HttpStatus.GONE);
```
Comment (gsergiu) [inline apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java]: the response should also indicate the reason. (e.g. api key is deprecated)

File: apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java [hunk 4]
```diff
@@ -313,61 +315,29 @@ public ApikeyController(ApikeyRepo apikeyRepo) {
...
-                int oops = 0 / 0;
-            } catch (ArithmeticException e) {
-                LOG.error("Deliberate error thrown: ", e);
-            }
-        } else if (!method.equalsIgnoreCase(READ) && !method.equalsIgnoreCase(WRITE)) {
-                return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
         }
 
+        LOG.debug("validate apikey: {}", id);
+
```
Comment (gsergiu) [inline apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java]: I assume that the following line might trow a runtime exception that is not catched within this method....

File: apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java [hunk 5]
```diff
@@ -386,7 +358,22 @@ public ApikeyController(ApikeyRepo apikeyRepo) {
         this.apikeyRepo.save(apikey);
```
Comment (gsergiu) [inline apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java]: I assume that this method might throw a runtime exception that is not handled....

File: apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java [hunk 6]
```diff
@@ -386,7 +358,22 @@ public ApikeyController(ApikeyRepo apikeyRepo) {
...
+    private String getApikey(HttpServletRequest httpServletRequest) {
+        String authorization = httpServletRequest.getHeader("Authorization");
+        if (authorization != null) {
+            String patternString = APIKEY_PATTERN;
+
+            Pattern pattern = Pattern.compile(patternString);
+            Matcher matcher = pattern.matcher(authorization);
+
+            if (matcher.find()) {
+                return matcher.group(1);
```
Comment (gsergiu) [inline apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java]: shoudn't group() be sufficient? the group() methods throw runtime exceptions that are not handled ...

File: apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java (hunk 4 above)
Comment (helinek) [reply apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java]: Why do you assume that?

File: apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java (hunk 6 above)
Comment (helinek) [reply apikey-service/src/main/java/eu/europeana/apikey/controller/ApikeyController.java]: No, group() is not sufficient, it calls group(0) while here is group(1).

**Pull request merged**